yarn-project/p2p/src/client/p2p_client.ts

```diff
@@ -558,7 +602,37 @@ export class P2PClient extends WithTracer implements P2P {
     await this.startServiceIfSynched();
   }
 
+  /**
+   * Updates the tx pool after a chain prune.
+   * @param latestBlock - The block number the chain was pruned to.
+   */
+  private async handlePruneL2Blocks(latestBlock: number): Promise<void> {
+    const txsToDelete: TxHash[] = [];
+    for (const tx of this.txPool.getAllTxs()) {
+      // every tx that's been generated against a block that has now been pruned is no longer valid
+      // NOTE (alexg): I think this check against block hash instead of block number?
+      if (tx.data.constants.globalVariables.blockNumber.toNumber() > latestBlock) {
+        txsToDelete.push(tx.getTxHash());
+      }
+    }
+
+    // TODO (alexg): Delete or re-add txs that were created against the proven block but mined in one of the pruned blocks
+    // e.g. I create a tx against proven block 42 but it the sequencer includes it in block 45. The chain gets pruned back to 42.
+    // That tx now lingers in the pool as 'mined' but it really is no longer mined. It's also not technically invalid.
+
+    this.log.info(
+      `Detected chain prune. Removing invalid txs count=${
+        txsToDelete.length
+      } newLatestBlock=${latestBlock} previousLatestBlock=${this.getSyncedLatestBlockNum()}`,
+    );
+    await this.txPool.deleteTxs(txsToDelete);
+    await this.synchedLatestBlockNumber.set(latestBlock);
+    await this.synchedProvenBlockNumber.set(latestBlock);
```

Comment:
I'm not sure we should be setting the synched proven block number here. This should only be done via proven chain events shouldn't it?

Reply:
I was thinking of the case where the proven chain is rolled back due to finalisation on L1 (which would still be emitted by `L2BlocKStream` as a `chain-pruned` event).

Reply:
Removed :)